kratos/elements/levelset_convection_element_simplex_algebraic_stabilization.h

```diff
@@ -0,0 +1,434 @@
+
+        BoundedMatrix<double,TNumNodes, TNumNodes> Ncontainer;
+        GetShapeFunctionsOnGauss(Ncontainer);
+        for(unsigned int igauss=0; igauss<TDim+1; igauss++)
+        {
+            noalias(N) = row(Ncontainer,igauss);
+
+            //obtain the velocity/coordinate at the gauss point
+            array_1d<double, TDim > vel_gauss=ZeroVector(TDim);
+            array_1d<double, TDim > X_gauss=ZeroVector(TDim);
+            double phi_gauss = 0.0;
+            double phi_gauss_old = 0.0;
+            for (unsigned int i = 0; i < TNumNodes; i++)
+            {
+                for(unsigned int k=0; k<TDim; k++)
+                {
+                    vel_gauss[k] += N[i]*v[i][k];
+                    X_gauss[k] += N[i]*X_node[i][k];
+                }
+                phi_gauss += N[i]*phi[i];
+                phi_gauss_old += N[i]*phi_old[i];
+            }
+
+            array_1d<double, TNumNodes > v_dot_grad_N = prod(DN_DX, vel_gauss);
+
+            noalias(Mc_matrix) += outer_prod(N, N);
+            noalias(K_matrix) += outer_prod(N, v_dot_grad_N);
+
+            for (unsigned int i = 0; i < TNumNodes; i++){
+                S_vector[i] += ( (phi_gauss_old - phi_mean_old) - inner_prod(grad_phi_mean,(X_gauss - X_mean)) )*N[i];
+            }
+        }
+
+        for (unsigned int i = 0; i < TNumNodes; i++)
+        {
+            Ml_matrix(i,i) = 1.0;
+        }
+
+        noalias(S_matrix) = (1.0/static_cast<double>(TNumNodes))*(Ml_matrix-Mc_matrix);
+
+        double nu_e = 0.0;
+        for (unsigned int i = 0; i < TNumNodes; i++)
+        {
+            for (unsigned int j = 0; j < TNumNodes; j++)
+            {
+                if (i == j)
+                    continue;
+
+                nu_e = std::max( nu_e, std::max(0.0, K_matrix(i, j))/std::abs(S_matrix(i, j)) );
+            }
+        }
+
+        const double limiter = GetValue(LIMITER_COEFFICIENT);
+        noalias(rLeftHandSideMatrix)  = dt_inv*((1.0-limiter)*Ml_matrix + limiter*Mc_matrix) + theta*(K_matrix + (1.0-limiter)*nu_e*S_matrix);
+        noalias(rRightHandSideVector) = prod( dt_inv*((1.0-limiter)*Ml_matrix + limiter*Mc_matrix) - (1.0 - theta)*(K_matrix + (1.0-limiter)*nu_e*S_matrix) , phi_old) - limiter*nu_e*S_vector;
+
+        //take out the dirichlet part to finish computing the residual
+        noalias(rRightHandSideVector) -= prod(rLeftHandSideMatrix, phi);
+
+        rRightHandSideVector *= Volume/static_cast<double>(TNumNodes);
+        rLeftHandSideMatrix *= Volume/static_cast<double>(TNumNodes);
+
+        KRATOS_CATCH("Error in Levelset Element")
+    }
+
+    void CalculateRightHandSide(VectorType& rRightHandSideVector, const ProcessInfo& rCurrentProcessInfo) override
+    {
+        KRATOS_THROW_ERROR(std::runtime_error, "CalculateRightHandSide not implemented","");
+    }
+
+    void EquationIdVector(EquationIdVectorType& rResult, const ProcessInfo& rCurrentProcessInfo) const override
```

Comment:
This is identical in the other levelset convection element. Why not deriving from it?

Reply:
I had never thought of that.